Preliminary support for custom compression algorithms

pdillinger opened the pull request:
Summary: This change builds on #13540 and #13626 in allowing a CompressionManager / Compressor / Decompressor to use a custom compression algorithm, with a distinct CompressionType. For background, review the API comments on CompressionManager and its CompatibilityName() function.

Highlights:
* Reserve and name 127 new CompressionTypes that can be used for custom compression algorithms / schemas. In many or most cases I expect the enumerators such as `kCustomCompression8F` to be used in user code rather than casting between integers and CompressionTypes, as I expect the supported custom compression algorithms to be identifiable / enumerable at compile time.
* When using these custom compression types, a CompressionManager must use a CompatibilityName() other than the built-in one AND new format_version=7 (see below).
* When building new SST files, track the full set of CompressionTypes actually used (usually just one aside from kNoCompression), using our efficient bitset SmallEnumSet, which supports fast iteration over the bits set to 1. Ideally, to support mixed or non-mixed compression algorithms in a file as efficiently as possible, we would know the set of CompressionTypes as SST file open time.
* New schema for `TableProperties::compression_name` in format_version=7 to represent the CompressionManager's CompatibilityName(), the set of CompressionTypes used, and potentially more in the future, while keeping the data relatively human-readable.
  * It would be possible to do this without a new format_version, but then the only way to ensure incompatible versions fail is with an unsupported CompressionType tag, not with a compression_name property. Therefore, (a) I prefer not to put something misleading in the `compression_name` property (a built-in compression name) when there is nuance because of a CompressionManager, and (b) I prefer better, more consistent error messages that refer to either format_version or the CompressionManager's CompatibilityName(), rather than an unrecognized custom CompressionType value (which could have come from various CompressionManagers).
* The current configured CompressionManager is passed in to TableReaders so that it (or one it knows about) can be used if it matches the CompatibilityName() used for compression in the SST file. Until the connection with ObjectRegistry is implemented, the only way to read files generated with a particular CompressionManager using custom compression algorithms is to configure it (or a known relative; see FindCompatibleCompressionManager()) in the ColumnFamilyOptions.
* Optimized snappy compression with BuiltinDecompressorV2SnappyOnly, to offset some small added overheads with the new tracking. This is essentially an early part of the planned refactoring that will get rid of the old internal compression APIs.
* Another small optimization in eliminating an unnecessary key copy in flush (builder.cc).
* Fix some handling of named CompressionManagers in CompressionManager::CreateFromString() (problem seen in #13647)

Smaller things:
* Adds Name() and GetId() functions to Compressor for debugging/logging purposes. (Compressor and Decompressor are not expected to be Customizable because they are only instantiated by a CompressionManager.)
* When using an explicit compression_manager, the GetId() of the CompressionManager and the Compressor used to build the file are stored as bonus entries in the compression_options table property. This table property is not parsed anywhere, so it is currently for human reading, but still could be parsed with the new underscore-prefixed bonus entries. IMHO, this is preferable to additional table properties, which would increase memory fragmentation in the TableProperties objects and likely take slightly more CPU on SST open and slightly more storage.
* ReleaseWorkingArea() function from protected to public to make wrappers work, because of a quirk in C++ (vs. Java) in which you cannot access protected members of another instance of the same class (sigh)
* Added `CompressionManager:: SupportsCompressionType()` for early options sanity checking.

Follow-up before release:
* Make format_version=7 official / supported
* Stress test coverage

Sooner than later:
* Update tests for RoundRobinManager and SimpleMixedCompressionManager to take advantage of e.g. set of compression types in compression_name property
* ObjectRegistry stuff
* Refactor away old internal compression APIs

Test Plan: Basic unit test added.

## Performance

### SST write performance
```
SUFFIX=`tty | sed 's|/|_|g'`; for ARGS in "-compression_type=none" "-compression_type=snappy" "-compression_type=zstd" "-compression_type=snappy -verify_compression=1" "-compression_type=zstd -verify_compression=1" "-compression_type=zstd -compression_max_dict_bytes=8180"; do echo $ARGS; (for I in `seq 1 20`; do BIN=/dev/shm/dbbench${SUFFIX}.bin; rm -f $BIN; cp db_bench $BIN; $BIN -db=/dev/shm/dbbench$SUFFIX --benchmarks=fillseq -num=10000000 -compaction_style=2 -fifo_compaction_max_table_files_size_mb=1000 -fifo_compaction_allow_compaction=0 -disable_wal -write_buffer_size=12000000 -format_version=7 $ARGS 2>&1 | grep micros/op; done) | awk '{n++; sum += $5;} END { print int(sum / n); }'; done
```

Ops/sec, Before -> After, both fv=6:
-compression_type=none
1894386 -> 1858403 (-2.0%)
-compression_type=snappy
1859131 -> 1807469 (-2.8%)
-compression_type=zstd
1191428 -> 1214374 (+1.9%)
-compression_type=snappy -verify_compression=1
1861819 -> 1858342 (+0.2%)
-compression_type=zstd -verify_compression=1
979435 -> 995870 (+1.6%)
-compression_type=zstd -compression_max_dict_bytes=8180
905349 -> 940563 (+3.9%)

Ops/sec, Before fv=6 -> After fv=7:
-compression_type=none
1879365 -> 1836159 (-2.3%)
-compression_type=snappy
1865460 -> 1830916 (-1.9%)
-compression_type=zstd
1191428 -> 1210260 (+1.6%)
-compression_type=snappy -verify_compression=1
1866756 -> 1818989 (-2.6%)
-compression_type=zstd -verify_compression=1
982640 -> 997129 (+1.5%)
-compression_type=zstd -compression_max_dict_bytes=8180
912608 -> 937248 (+2.7%)

### SST read performance
Create DBs
```
for COMP in none snappy zstd; do echo $ARGS; ./db_bench -db=/dev/shm/dbbench-7-$COMP --benchmarks=fillseq,flush -num=10000000 -compaction_style=2 -fifo_compaction_max_table_files_size_mb=1000 -fifo_compaction_allow_compaction=0 -disable_wal -write_buffer_size=12000000 -compression_type=$COMP -format_version=7; done
```
And test
```
for COMP in none 
snappy zstd none; do echo $COMP; (for I in `seq 1 8`; do ./db_bench -readonly -db=/dev/shm/dbbench
-7-$COMP --benchmarks=readrandom -num=10000000 -duration=20 -threads=8 2>&1 | grep micros/op; done
) | awk '{n++; sum += $5;} END { print int(sum / n); }'; done
```

Ops/sec, Before -> After (both fv=6)
none
1491732 -> 1500209 (+0.6%)
snappy
1157216 -> 1169202 (+1.0%)
zstd
695414 -> 703719 (+1.2%)
none (again)
1491787 -> 1528789 (+2.4%)

Ops/sec, Before fv=6 -> After fv=7:
none
1492278 -> 1508668 (+1.1%)
snappy
1140769 -> 1152613 (+1.0%)
zstd
696437 -> 696511 (+0.0%)
none (again)
1500585 -> 1512037 (+0.7%)

Overall, I think we can take the read CPU improvement in exchange for the hit (in some cases) on background write CPU

hx235 commented on db/db_test2.cc:
```diff
@@ -2103,6 +2104,254 @@ TEST_F(DBTest2, CompressionManagerWrapper) {
+  EXPECT_EQ(mgr_foo->used_compressor8A_count_, 0);
+  ASSERT_OK(Flush());
+  ASSERT_EQ(NumTableFilesAtLevel(0), 3);
+  ASSERT_NE(Get("c"), "NOT_FOUND");
```
Can we also assert the value is correct too just to ensure no wrong decompression? Same for other Get() calls

pdillinger replied:
Done

hx235 commented on db/db_test2.cc:
```diff
@@ -2103,6 +2104,254 @@ TEST_F(DBTest2, CompressionManagerWrapper) {
+  ASSERT_EQ(tables_properties.size(), 1U);
+  EXPECT_LT(tables_properties.begin()->second->data_size, kValueSize / 2);
+  // Uses new format for "compression_name" property
+  EXPECT_EQ(tables_properties.begin()->second->compression_name, "Foo;8A;");
```
The list of CompressionTypes is nice. Is it possible to do something like that for the simple mixed compression unit test to check storing and retrieving a list is correct?

pdillinger replied:
I think there are potentially substantial updates there, so I prefer to defer that. (New follow-up item listed.)

hx235 commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -954,6 +970,92 @@ struct BlockBasedTableBuilder::Rep {
+          compression_name = CompressionTypeToString(
+              data_block_compressor->GetPreferredCompressionType());
+        } else {
+          assert(basic_compressor == nullptr);
```
is it possible that data_block_compressor is nullptr because of below returns a nullptr while basic_compressor is initialized? 

```
  r->compressor_with_dict = r->basic_compressor->MaybeCloneSpecialized(
      CacheEntryRole::kDataBlock, std::move(samples));
 ```

pdillinger replied:
Yes, EnterUnbuffered() always sets the data_block_compressor (if basic_compressor is set)

hx235 commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -770,15 +776,25 @@ struct BlockBasedTableBuilder::Rep {
+                .get();
+    } else {
+      uses_explicit_compression_manager = true;
+
+      // Stuff some extra debugging info as extra pseudo-options. Using
+      // underscore prefix to indicate they are special.
```
nit: maybe explicitly add words like "debug_info"?

hx235 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -562,6 +562,102 @@ Status GetGlobalSequenceNumber(const TableProperties& table_properties,
+      if (!s.ok()) {
+        return s;
+      }
+      assert(mgr_to_use);
```
Can be moved before "// Second field is set of compression types actually used in the file" ?

pdillinger replied:
If the assertion fails, you get more information where it is, because you know you were under the `if (mgr_to_use == nullptr)`. Moving it outside is basically redundant with the later deref.

hx235 commented:
Left some comments that should be easy to fix. Look forward to the stress test coverage! Are regression for compression_type=none and snappy noise or concerning? If it's just noise, I can approve the PR now (assuming the stress test coverage will happen soon)

pdillinger commented:
Thanks! @hx235 

> @pdillinger Left some comments that should be easy to fix. Look forward to the stress test coverage! Are regression for compression_type=none and snappy noise or concerning? If it's just noise, I can approve the PR now (assuming the stress test coverage will happen soon)

Updated performance conclusion: "Overall, I think we can take the read CPU improvement in exchange for the hit (in some cases) on background write CPU."

I've been looking at the write benchmark closely and it's very repeatable. I added a part of my test script that copies the test binary to a new file for each run. I was seeing cases in which the copies of the same binary could have dramatically different performance (~10%), repeatedly. I don't quite know what's up with that, perhaps some physical memory fragmentation or something, but I think that's factored out with this extra step.

The read benchmark has some noise, but I don't think we need to dig too much since it's pretty consistently slightly good in favor of the changes.

pdillinger commented on db/db_test2.cc:
```diff
@@ -2103,6 +2104,254 @@ TEST_F(DBTest2, CompressionManagerWrapper) {
+  EXPECT_EQ(tables_properties.begin()->second->compression_name, "Snappy");
+
+  // TODO: check what happens trying to set a custom CompressionType with a
+  // built-in compatible CompressionManager
```
Done
